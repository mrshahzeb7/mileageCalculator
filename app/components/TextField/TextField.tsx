import React, {ComponentType, Ref, forwardRef, useRef} from 'react';
import {
  StyleProp,
  StyleSheet,
  Text,
  TextInput,
  TextInputProps as TextInputPropsType,
  TextStyle,
  View,
  ViewStyle,
} from 'react-native';
import {borderRadius, colors, fontSize, sizes, spacing} from '../../theme';
import {responsiveHeight} from '../../utils';
import {Label} from '../Label/Label';

export interface ValidatedInputType {
  isValid: boolean;
  value: string;
}

export type onChangeValidatedInputType = (data: ValidatedInputType) => void;

export interface TextFieldProps extends Omit<TextInputPropsType, 'ref'> {
  label?: string;
  placeholder?: string;
  style?: StyleProp<TextStyle>;
  containerStyle?: StyleProp<ViewStyle>;
  inputWrapperStyle?: StyleProp<ViewStyle>;
  RightAccessory?: ComponentType<any>;
  LeftAccessory?: ComponentType<any>;
  helper?: string;
}

export const TextField = forwardRef(function TextField(
  props: TextFieldProps,
  ref: Ref<TextInput>,
) {
  const {
    label,
    placeholder,
    RightAccessory,
    LeftAccessory,
    style: $inputStyleOverride,
    containerStyle: $containerStyleOverride,
    inputWrapperStyle: $inputWrapperStyleOverride,
    helper,
    ...TextInputProps
  } = props;

  const disabled = TextInputProps.editable === false;

  const $containerStyles = [$containerStyleOverride];

  const $inputWrapperStyles = [
    styles.inputWrapperStyle,
    LeftAccessory && {paddingStart: 0},
    RightAccessory && {paddingEnd: 0},
    $inputWrapperStyleOverride,
  ];

  const $inputStyles: StyleProp<ViewStyle> = [
    styles.inputStyle,
    TextInputProps.multiline && {height: 'auto'},
    $inputStyleOverride,
  ];

  return (
    <View style={$containerStyles}>
      <Label label={label} />

      <View style={$inputWrapperStyles}>
        {!!LeftAccessory && <LeftAccessory />}

        <TextInput
          {...(disabled ? {pointerEvents: 'none'} : null)}
          ref={ref}
          underlineColorAndroid={colors.transparent}
          textAlignVertical="top"
          placeholder={placeholder}
          placeholderTextColor={colors.textDim}
          maxLength={props.maxLength || 30}
          {...TextInputProps}
          editable={!disabled}
          style={$inputStyles}
        />

        {!!RightAccessory && <RightAccessory />}
      </View>
      {!!helper && <Text style={styles.helper}>{helper}</Text>}
    </View>
  );
});

export function useTextFieldRef() {
  const textFieldRef = useRef<TextInput>(null);
  return textFieldRef;
}

const styles = StyleSheet.create({
  inputWrapperStyle: {
    flexDirection: 'row',
    alignItems: 'center',
    borderWidth: responsiveHeight(0.1),
    height: sizes.medium,
    borderColor: colors.border,
    overflow: 'hidden',
    borderRadius: borderRadius.small,
  },

  inputStyle: {
    flex: 1,
    alignSelf: 'stretch',
    color: colors.text,
    fontSize: fontSize.medium,
    paddingVertical: 0,
    paddingHorizontal: 0,
    marginHorizontal: spacing.horizontal.medium,
    textAlignVertical: 'center',
  },
  helper: {
    marginLeft: spacing.horizontal.medium,
    fontSize: fontSize.small,
    backgroundColor: colors.primary,
    paddingHorizontal: spacing.horizontal.tiny,
    color: colors.textDim,
  },
});
