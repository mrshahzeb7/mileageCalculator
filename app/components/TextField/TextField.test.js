import 'react-native';
import React from 'react';
import {TextField} from './TextField';
import renderer from 'react-test-renderer';

describe('TextField', () => {
  it(' TextField renders correctly', () => {
    const tree = renderer.create(<TextField />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
