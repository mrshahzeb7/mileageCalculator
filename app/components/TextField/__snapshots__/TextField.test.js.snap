// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField  TextField renders correctly 1`] = `
<View
  style={
    [
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderColor": "#6a6f71",
          "borderRadius": 15.297966458904849,
          "borderWidth": 1.334,
          "flexDirection": "row",
          "height": 80.03999999999999,
          "overflow": "hidden",
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    <TextInput
      editable={true}
      maxLength={30}
      placeholderTextColor="#6a6f71"
      style={
        [
          {
            "alignSelf": "stretch",
            "color": "#ffffff",
            "flex": 1,
            "fontSize": 27.53633962602873,
            "marginHorizontal": 37.5,
            "paddingHorizontal": 0,
            "paddingVertical": 0,
            "textAlignVertical": "center",
          },
          undefined,
          undefined,
        ]
      }
      textAlignVertical="top"
      underlineColorAndroid="transparent"
    />
  </View>
</View>
`;
